dali/kernels/shape.h

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SHAPE_H_
+#define DALI_KERNELS_SHAPE_H_
+
+#include <dali/common.h>
+#include <vector>
+#include <array>
+#include <cassert>
+#include "util.h"
+
+namespace dali {
+
+template <typename T>
+constexpr typename std::enable_if<std::is_fundamental<T>::value, size_t>::type ShapeDim(const T &)
+{ return 1; }
+
+template <typename T, size_t N>
+constexpr int ShapeDim(T (&)[N]) { return int(N); }
```

Comment:
any reason for not returning size_t?

Reply:
Negative number of dimensions means "dynamic". But it's more related to https://github.com/NVIDIA/DALI/pull/346/files than this PR, actually.